crates/rattler_index/src/lib.rs

```diff
@@ -97,117 +127,317 @@ pub fn package_record_from_conda(file: &Path) -> std::io::Result<PackageRecord>
+/// If `target_platform` is `Some`, only that specific subdir is indexed.
+/// Otherwise indexes all subdirs and creates a `repodata.json` for each.
+///
+/// The process is the following:
+/// 1. Get all subdirs and create `noarch` and `target_platform` if they do not exist.
+/// 2. Iterate subdirs and index each subdir.
+///    Therefore, we need to:
+///    1. Collect all uploaded packages in subdir
+///    2. Collect all registered packages from `repodata.json` (if exists)
+///    3. Determine which packages to add to and to delete from `repodata.json`
+///    4. Write `repodata.json` back to disk
+pub async fn index<T: Configurator>(
+    target_platform: Option<Platform>,
+    config: T,
+    force: bool,
+    max_parallel: usize,
+) -> anyhow::Result<()> {
+    let builder = config.into_builder();
+
+    // Get all subdirs
+    let op = Operator::new(builder)?.finish();
+    let entries = op.list_with("").await?;
+    let mut subdirs = entries
+        .iter()
+        .filter_map(|entry| {
+            if entry.metadata().mode().is_dir() && entry.name() != "/" {
+                // Directory entries always end with `/`.
+                Some(entry.name().trim_end_matches('/').to_string())
+            } else {
+                None
+            }
+        })
+        .map(|s| Platform::from_str(&s))
+        .collect::<Result<HashSet<_>, _>>()?;
 
-        let mut out_file =
-            NamedTempFile::with_prefix_in("repodata-", output_folder.join(&platform))?;
-        out_file.write_all(serde_json::to_string_pretty(&repodata)?.as_bytes())?;
-        out_file.persist(output_folder.join(&platform).join("repodata.json"))?;
+    // If requested `target_platform` subdir does not exist, we create it.
+    if let Some(target_platform) = target_platform {
+        tracing::debug!("Did not find {target_platform} subdir, creating.");
+        if !subdirs.contains(&target_platform) {
+            op.create_dir(&format!("{}/", target_platform.as_str()))
+                .await?;
+        }
+        // Limit subdirs to only the requested `target_platform`.
+        subdirs = HashSet::default();
+        subdirs.insert(target_platform);
+    } else if !subdirs.contains(&Platform::NoArch) {
+        // If `noarch` subdir does not exist, we create it.
+        tracing::debug!("Did not find noarch subdir, creating.");
+        op.create_dir(&format!("{}/", Platform::NoArch.as_str()))
+            .await?;
+        subdirs.insert(Platform::NoArch);
     }
 
+    let multi_progress = MultiProgress::new();
+    let semaphore = Semaphore::new(max_parallel);
+    let semaphore = Arc::new(semaphore);
+
+    let tasks = subdirs
+        .iter()
+        .map(|subdir| {
+            tokio::spawn(index_subdir(
+                *subdir,
+                op.clone(),
+                force,
+                multi_progress.clone(),
+                semaphore.clone(),
+            ))
+        })
+        .collect::<Vec<_>>();
+    try_join_all(tasks).await?;
```

Comment:
we should add a `println!("Finished indexing ...")` here as otherwise we won't have any output in non-tty environments